### PR TITLE
gRPC benchmark apps

### PR DESCRIPTION
**Pull request by JamesNK**
BenchmarkServer: A Grpc.AspNetCore.Server, usable from the commandline and the benchmark envirnoment

NativeServer: A Grpc.Core native bits server, usable from the commandline and the benchmark envirnoment

BenchmarkClient: A console app for making requests from .NET using the Grpc.Core native bits

### CODE REVIEW

File: perf/BenchmarkServer/BenchmarkServer.csproj [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    <!-- 3. None ItemGroup: Remove the *.proto files from None since the generator metadata for those files are ignored -->
+    <!-- 4. Content ItemGroup: Include the *.proto files in the Content ItemGroup. This isn't required if the file is included in the project directory. I'm not sure about what's going on with this requirement. -->
+    <ProtoBuf Include="greet.proto" GrpcServices="Server" Generator="MSBuild:Compile" />
+
+    <None Remove="@(ProtoBuf)" />
+    <Content Include="@(ProtoBuf)" LinkBase="" />
+
+    <Compile Include="$(MSBuildThisFileDirectory)..\..\Sample\Shared\Resources.cs" />
+
+    <ProjectReference Include="../../src/Grpc.AspNetCore/Grpc.AspNetCore.csproj" />
```
Comment (sebastienros) [inline perf/BenchmarkServer/BenchmarkServer.csproj]: Do we generate myget packages for it? If so we might want to do a package reference later on when we make it a recurring job on the CI. Only in the case of `BenchmarksTargetFramework != ''`

**Comment by sebastienros** [comment]
It's confusing me that you called it `BenchmarksServer`, because it could be confused with our real `BenchmarksServer`.

File: perf/BenchmarkServer/Greeter.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        _logger.LogInformation($"Sending hello to {request.Name}");
+        return Task.FromResult(new HelloReply
+        {
+            Message = "Hello " + request.Name,
+            Timestamp = Timestamp.FromDateTime(DateTime.UtcNow)
+        });
+    }
+
+    public override async Task SayHellos(HelloRequest request, IServerStreamWriter<HelloReply> responseStream, ServerCallContext context)
```
Comment (JunTaoLuo) [inline perf/BenchmarkServer/Greeter.cs]: I don't think this streaming call will be needed right? Even if we are setting up streaming, I think we should be streaming something more reasonable.

File: perf/BenchmarkServer/Program.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            WebHost.CreateDefaultBuilder(args)
+                .ConfigureKestrel((context, options) =>
+                {
+                    options.Limits.MinRequestBodyDataRate = null;
+
+                    var endPoint = context.Configuration.CreateIPEndPoint();
+
+                    options.Listen(endPoint, listenOptions =>
+                    {
+                        listenOptions.UseHttps(Resources.ServerPFXPath, "1111");
```
Comment (JunTaoLuo) [inline perf/BenchmarkServer/Program.cs]: I created my own set of certs for a proof of concept, is it okay to be using them here?

Comment (JamesNK) [reply perf/BenchmarkServer/Program.cs]: I don't see why not. What are you worried about?

File: perf/BenchmarkServer/Greeter.cs (hunk 2 above)
Comment (JamesNK) [reply perf/BenchmarkServer/Greeter.cs]: Its not used. Once we have this passing I'll include benchmark calls that are similar to what grpc/grpc does for benchmarking.

File: perf/BenchmarkServer/BenchmarkServer.csproj (hunk 1 above)
Comment (JamesNK) [reply perf/BenchmarkServer/BenchmarkServer.csproj]: No NuGet packages yet. This is very early.

**Comment by JunTaoLuo** [comment]
cc @jtattermusch

**Comment by JamesNK** [comment]
The website and client in this PR are focused on the ASP.NET performance environment. We want to setup gRPC benchmarks so we can test the performance of ASP.NET Core's HTTP/2 stack. Our HTTP/2 stack is very new and gRPC will be one of the first real-world samples. Having it to test with will be vital to making it fast.

The service calls will definitely change to be in line with https://github.com/grpc/grpc/blob/master/src/proto/grpc/testing/benchmark_service.proto. Right now I just want to get it working with something.

I think the work to put AspNetCore.Server in the gRPC's benchmarks will happen separately. Is it something you could lead?

File: perf/BenchmarkServer/Startup.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddGrpc();
+        }
+
+        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
+        {
+            app.UseEndpointRouting(builder =>
```
Comment (davidfowl) [inline perf/BenchmarkServer/Startup.cs]: nit: builder => routes

**Comment by JunTaoLuo** [comment]
Changes look good, but did we ever find out how to configure streams per connection and connections per client?

**Comment by JunTaoLuo** [comment]
Note, this will require migration to grpc/grpc

**Comment by jtattermusch** [comment]
@JamesNK  do you want do close this PR? Superseded by https://github.com/grpc/grpc-dotnet/pull/67

**Comment by JamesNK** [comment]
Sure!

**Pull request closed without merging**